internal/jwks/jwks.go

```diff
@@ -248,18 +251,41 @@ func (mgr *Manager) UpdateAll(ctx context.Context) error {
 	defer logger.Debug("finished jwks update")
 
 	haDB := hadb.New(mgr.db)
-	healthAuthorities, err := haDB.ListAllHealthAuthoritiesWithoutKeys(ctx)
+	allHealthAuthorities, err := haDB.ListAllHealthAuthoritiesWithoutKeys(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to query db: %w", err)
 	}
 
+	// Pre-filter out health authorities where discovery isn't enabled.
+	healthAuthorities := make([]*model.HealthAuthority, 0, len(allHealthAuthorities))
+	for _, ha := range allHealthAuthorities {
+		if ha.JWKS() {
+			healthAuthorities = append(healthAuthorities, ha)
+		}
+	}
+
+	// Crypto secure random isn't necessary here, just shuffling the scheduling order on each run.
+	rand.Shuffle(len(healthAuthorities), func(i, j int) {
+		healthAuthorities[i], healthAuthorities[j] = healthAuthorities[j], healthAuthorities[i]
+	})
+
 	var merr *multierror.Error
 	var merrLock sync.Mutex
-
+	sem := semaphore.NewWeighted(int64(mgr.maxWorkers))
 	var wg sync.WaitGroup
 	for _, ha := range healthAuthorities {
+		// Rate limit the number of concurrent workers with a semaphore.
+		if err := sem.Acquire(ctx, 1); err != nil {
+			logger.Errorw("failed to acquire semaphore", "error", err)
+			merrLock.Lock()
+			merr = multierror.Append(merr, fmt.Errorf("failed to processes %v: %w", ha.Name, err))
+			merrLock.Unlock()
+			break
```

Comment:
continue?

Reply:
no - if we can't get the semaphore, it's almost certainly because the context timed out